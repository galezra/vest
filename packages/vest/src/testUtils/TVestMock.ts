import { TFieldName, TGroupName } from 'SuiteResultTypes';
import * as vest from 'vest';

export type TVestMock = typeof vest;

export type TTestSuiteCallback = (..._args: any[]) => void;
export type TTestSuite = vest.Suite<TFieldName, TGroupName, TTestSuiteCallback>;
