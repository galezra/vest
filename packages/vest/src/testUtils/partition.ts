export default function partition<T>(
  array: T[],
  predicate: (value: T, _index: number, array: T[]) => boolean,
): [T[], T[]] {
  return array.reduce(
    (partitions: [T[], T[]], value, number) => {
      partitions[predicate(value, number, array) ? 0 : 1].push(value);
      return partitions;
    },
    [[], []],
  );
}
