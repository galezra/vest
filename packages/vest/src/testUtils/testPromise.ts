export function TestPromise(cb: (_done: () => void) => void): Promise<void> {
  return new Promise<void>(done => cb(done));
}
