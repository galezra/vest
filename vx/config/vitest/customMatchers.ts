import { isFunction } from 'vest-utils';

expect.extend({
  isDeepCopyOf,
});

// eslint-disable-next-line complexity
function isDeepCopyOf(
  source: any,
  clone: any,
): { pass: boolean; message: () => string } {
  const queue = [[source, clone]];

  outer: while (queue.length) {
    const [source, clone] = (queue as any[]).shift();

    if (!source || typeof source !== 'object') {
      if (!isFunction(clone)) {
        if (clone !== source) {
          return {
            pass: false,
            message: () => `Source and clone are not identical`,
          };
        }
      }
      continue;
    }

    if (clone === source) {
      return {
        pass: false,
        message: () =>
          `Source and clone are the same object. Expected a deep copy. ${JSON.stringify(
            source,
          )}===${JSON.stringify(clone)}`,
      };
    }

    // Short circuit
    if (
      (clone && !source) ||
      (source && !clone) ||
      typeof source !== typeof clone
    ) {
      return {
        pass: false,
        message: () =>
          `Source and clone are not of the same type: ${JSON.stringify(
            source,
          )} does not equal ${JSON.stringify(clone)}`,
      };
    }

    if (Array.isArray(source)) {
      // Short circuit
      if (!Array.isArray(clone) || source.length !== clone.length) {
        return {
          pass: false,
          message: () =>
            `source and clone arrays are not identical. ${JSON.stringify(
              source,
            )} does not equal ${JSON.stringify(clone)}`,
        };
      }
      source.forEach((_, i) => {
        queue.push([source[i], clone[i]]);
      });
    } else if (typeof source === 'object') {
      Object.keys(source).forEach(key => queue.push([source[key], clone[key]]));
    }

    continue outer;
  }

  return { pass: true, message: () => 'success' };
}
