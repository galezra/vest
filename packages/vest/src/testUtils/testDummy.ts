import { faker } from '@faker-js/faker';
import { vi } from 'vitest';

import { test as vestTest, warn } from 'vest';

/**
 * Generates dummy vest tests.
 */
// eslint-disable-next-line max-lines-per-function
const testDummy = () => {
  const failing = (
    name: string = faker.lorem.word(),
    message: string = faker.lorem.words(),
  ) => {
    const to = vestTest(
      name,
      message,
      vi.fn(() => {
        throw new Error();
      }),
    );

    return to;
  };

  const failingWarning = (
    name = faker.lorem.word(),
    message = faker.lorem.words(),
  ) => {
    const to = vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
        throw new Error();
      }),
    );

    return to;
  };

  const passing = (
    name = faker.lorem.word(),
    message = faker.lorem.words(),
  ) => {
    const to = vestTest(name, message, vi.fn());

    return to;
  };

  const passingWarning = (
    name = faker.lorem.word(),
    message = faker.lorem.words(),
  ) => {
    const to = vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
      }),
    );
    return to;
  };

  const failingAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(
        () =>
          new Promise((_, reject) => {
            setTimeout(reject, time);
          }),
      ),
    );

  const failingWarningAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
        return new Promise((_, reject) => {
          setTimeout(reject, time);
        });
      }),
    );

  const passingAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(
        () =>
          new Promise(resolve => {
            setTimeout(resolve, time);
          }),
      ),
    );

  const passingWarningAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
        return new Promise(resolve => {
          setTimeout(resolve, time);
        });
      }),
    );

  return {
    failing,
    failingAsync,
    failingWarning,
    failingWarningAsync,
    passing,
    passingAsync,
    passingWarning,
    passingWarningAsync,
  };
};

export const dummyTest = testDummy();

export type TDummyTest = typeof dummyTest;
