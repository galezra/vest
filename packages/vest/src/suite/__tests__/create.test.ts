import { faker } from '@faker-js/faker';
import { ErrorStrings } from 'ErrorStrings';
import { noop } from 'lodash';
import { describe, it, expect, vi } from 'vitest';

import { dummyTest } from '../../testUtils/testDummy';
import { TestPromise } from '../../testUtils/testPromise';

import { create } from 'vest';

describe('Test createSuite module', () => {
  describe('Test suite Arguments', () => {
    it('allows omitting suite name', () => {
      expect(typeof create(vi.fn())).toBe('function');
      expect(typeof create(vi.fn()).get).toBe('function');
      expect(typeof create(vi.fn()).reset).toBe('function');
      expect(typeof create(vi.fn()).remove).toBe('function');
      expect(create(vi.fn()).get()).toMatchSnapshot();
    });

    it.each([faker.lorem.word(), null, undefined, 0, 1, true, false, NaN, ''])(
      'Throws an error when `tests` callback is not a function',
      value => {
        // @ts-expect-error - testing invalid input
        expect(() => create(value)).toThrow(
          ErrorStrings.SUITE_MUST_BE_INITIALIZED_WITH_FUNCTION,
        );
      },
    );

    describe('When suite name is provided', () => {
      it('Should add suite name to suite result', () => {
        const res = create('form_name', () => {})();

        expect(res.suiteName).toBe('form_name');
      });
    });
  });

  describe('Return value', () => {
    it('should be a function', () => {
      expect(typeof create(noop)).toBe('function');
    });
  });

  describe('When returned function is invoked', () => {
    it('Calls `tests` argument', () =>
      TestPromise(done => {
        const validate = create(() => {
          done();
        });
        validate();
      }));

    it('Passes all arguments over to tests callback', () => {
      const testsCallback = vi.fn();
      const params = [
        1,
        2,
        3,
        { [faker.lorem.word()]: [1, 2, 3] },
        false,
        [faker.lorem.word()],
      ];
      const validate = create(testsCallback);
      validate(...params);
      expect(testsCallback).toHaveBeenCalledWith(...params);
    });
  });

  describe('Initial run', () => {
    const testsCb = vi.fn();
    const genValidate = () => create(testsCb);

    it('Should initialize with an empty result object', () => {
      const validate = genValidate();
      expect(Object.keys(validate.get().tests)).toHaveLength(0);
      expect(Object.keys(validate.get().groups)).toHaveLength(0);

      expect(validate.get().errorCount).toBe(0);
      expect(validate.get().warnCount).toBe(0);
      expect(validate.get().testCount).toBe(0);

      expect(validate.get()).toMatchSnapshot();
    });

    it('Should be able to get the suite from the result of createSuite', () => {
      const testsCb = vi.fn();
      expect(create(testsCb).get()).toMatchSnapshot();
    });

    it('Should be able to reset the suite from the result of createSuite', () => {
      const testSuite = create(() => {
        dummyTest.failing('f1', 'm1');
      });
      testSuite();
      expect(testSuite.get().hasErrors()).toBe(true);
      expect(testSuite.get().testCount).toBe(1);
      testSuite.reset();
      expect(testSuite.get().hasErrors()).toBe(false);
      expect(testSuite.get().testCount).toBe(0);
    });

    it('Should return without calling tests callback', () => {
      const validate = create(testsCb);
      expect(testsCb).not.toHaveBeenCalled();
      validate();
      expect(testsCb).toHaveBeenCalled();
    });
  });
});
