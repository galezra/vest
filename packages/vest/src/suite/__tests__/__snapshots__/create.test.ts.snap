// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test createSuite module > Initial run > Should be able to get the suite from the result of createSuite 1`] = `
SuiteSummary {
  "errorCount": 0,
  "errors": [],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getMessage": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isPending": [Function],
  "isTested": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 0,
  "tests": {},
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;

exports[`Test createSuite module > Initial run > Should initialize with an empty result object 1`] = `
SuiteSummary {
  "errorCount": 0,
  "errors": [],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getMessage": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isPending": [Function],
  "isTested": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 0,
  "tests": {},
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;

exports[`Test createSuite module > Test suite Arguments > allows omitting suite name 1`] = `
SuiteSummary {
  "errorCount": 0,
  "errors": [],
  "getError": [Function],
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getMessage": [Function],
  "getWarning": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isPending": [Function],
  "isTested": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "pendingCount": 0,
  "suiteName": undefined,
  "testCount": 0,
  "tests": {},
  "valid": false,
  "warnCount": 0,
  "warnings": [],
}
`;
